integration/tests/services/test_ws_subscriber.py

```diff
@@ -0,0 +1,223 @@
+        [
+            (["Event2(string,string)"], None, 2),
+            ([], ["text1"], 3),
+            (["Event2(string,string)"], ["text2"], 1),
+            (["Event2(string,string)", "Event3(string,string,string)"], ["text2", "text3", "text1", "text5"], 3,),
+            ([], None, 4),
+        ],
+    )
+    async def test_filter_log_by_topics(self, event_filter, arg_filter, log_count):
+        async with websockets.connect(TEST_STAND) as ws:
+            event_caller, _ = self.web3_client.deploy_and_get_contract(
+                "EventCaller", "0.8.12", self.sender_account
+            )
+
+            arg1, arg2, arg3 = ("text1", "text2", "text3")
+            topics = []
+            if event_filter is not None:
+                event_topics = []
+                for item in event_filter:
+                    event_topics.append("0x" + keccak(text=item).hex())
+                topics.append(event_topics)
+            if arg_filter is not None:
+                arg_topics = []
+                for item in arg_filter:
+                    arg_topics.append("0x" + keccak(text=item).hex())
+                topics.append(arg_topics)
+
+            optional_fields = {"address": event_caller.address}
+            if topics:
+                optional_fields["topics"] = topics
+            params = ["logs", optional_fields]
+
+            data = Subscribe(params)
+            await ws.send(json.dumps(data.__dict__))
+            r = await ws.recv()
+            response = json.loads(r, object_hook=lambda d: SimpleNamespace(**d))
+            subscription = response.result
+            assert len(subscription) > 0
+
+            tx = self.make_tx_object(self.sender_account.address)
+            instruction_tx = event_caller.functions.callEvent1(arg1).build_transaction(tx)
+            self.web3_client.send_transaction(self.sender_account, instruction_tx)
+
+            tx = self.make_tx_object(self.sender_account.address)
+            instruction_tx = event_caller.functions.callEvent2(arg1, arg2).build_transaction(
+                tx
+            )
+            self.web3_client.send_transaction(self.sender_account, instruction_tx)
+
+            tx = self.make_tx_object(self.sender_account.address)
+            instruction_tx = event_caller.functions.callEvent2(arg2, arg3).build_transaction(
+                tx
+            )
+            self.web3_client.send_transaction(self.sender_account, instruction_tx)
+
+            tx = self.make_tx_object(self.sender_account.address)
+            instruction_tx = event_caller.functions.callEvent3(
+                arg1, arg2, arg3
+            ).build_transaction(tx)
+            self.web3_client.send_transaction(self.sender_account, instruction_tx)
+
+            messages = await ws_receive_all_messages(ws)
+            assert (len(messages) == log_count), f"Expected {log_count} event logs, but found {len(messages)}"
+
+            is_event_topic_in_list = False
```

Comment:
try to optimize this, maybe split it to several tests or something

Reply:
Fixed